lib/modules/blockchain_process/proxy.js

```diff
@@ -63,7 +64,31 @@ const parseResponse = function (ipc, resBody) {
   }
 };
 
-exports.serve = function (ipc, host, port, ws) {
+exports.serve = async function (ipc, host, port, ws, origin) {
+  const _origin = origin ? origin.split(',')[0] : undefined;
+  const start = Date.now();
+
+  function awaitTarget() {
+    return new Promise(resolve => {
+      utils.pingEndpoint(
+        canonicalHost(host), port, ws ? 'ws': false, 'http', _origin, async (err) => {
+          if (!err || (Date.now() - start > 10000)) {
+            // if (Date.now() - start > 10000) {
+            //   console.warn('!!! TIMEOUT !!!');
+            // } else {
+            //   console.warn('!!! CONNECT !!!');
+            // }
+            return resolve();
+          }
+          // console.warn('!!! WAITING !!!');
+          await utils.timer(250).then(awaitTarget).then(resolve);
```

Comment:
Do you think there may be some sort of memory leak that could occur here if the endpoint **never** responds to the ping (and always errors, ie if geth bombs out in an `embark run`). If i'm reading this correctly, the promises would keep adding to the stack until there is one that finally resolves (a recursive pattern).

Reply:
At most it will stack for 40 promises — `10000 / 250 = 40` — and then it will timeout and they'll all get swept up in the next gc pass. When applying this approach more generally, care must be taken to not pile Promise instances on the heap w/o some kind of guard (timeout).

Reply:
Ah yes good point, nice work!